### PR TITLE
[tune] Initial track integration

### PR DESCRIPTION
**Pull request by noahgolmant**
Introduces a minimally invasive utility for logging experiment results. A broad requirement for this tool is that it should integrate seamlessly with Tune execution; see below examples.

Specifically, it integrates [track](https://github.com/richardliaw/track) as a submodule of tune to track experiment metrics.

Waiting on #4011.

#### Ray Version
```python
import ray
from ray.tune import track

ray.init()

def hello_world(config):
    print("hello world")
    track.metric(result="hello")

from ray import tune
tune.run(hello_world)
```

and 
#### Local Version
```python
from ray.tune import track


def hello_world(config):
    print("hello world")
    track.metric(result="hello")

track.init()
hello_world({})
```

Core features Needed:
 - Git tracking
 - Logging at iteration
 - Seamlessly flows into Tensorboard
 - Can sort results and config via `tune ls`

### CODE REVIEW

File: python/ray/tune/track/error.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+from __future__ import absolute_import
```
Comment (richardliaw) [inline python/ray/tune/track/error.py]: cut

File: python/ray/tune/track/convenience.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+"""
```
Comment (richardliaw) [inline python/ray/tune/track/convenience.py]: cut

File: python/ray/tune/track/autodetect.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
+"""
```
Comment (richardliaw) [inline python/ray/tune/track/autodetect.py]: cut for now, should be modularized from "track.trial"

File: python/ray/tune/track/trial.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+class _ReporterHook(Logger):
+    def __init__(self, reporter):
+        self.reporter = reporter
+
+    def on_result(self, metrics):
+        return self.reporter(**metrics)
+
+
+class Trial(object):
```
Comment (noahgolmant) [inline python/ray/tune/track/trial.py]: rename to `TrackContext` to distinguish from tune's `Trial`

File: python/ray/tune/track/log.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
+"""
```
Comment (richardliaw) [inline python/ray/tune/track/log.py]: cut (for now)

File: python/ray/tune/track/trial.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+        upload_dir (str):
+    """
+    def __init__(self,
+                 log_dir=None,
+                 upload_dir=None,
+                 sync_period=None,
+                 trial_prefix="",
+                 param_map=None,
+                 init_logging=True):
+        if log_dir is None:
```
Comment (noahgolmant) [inline python/ray/tune/track/trial.py]: Access tune's `UnifiedLogger` instead of handling logging locally

File: python/ray/tune/track/project.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
+try:  # py3
```
Comment (richardliaw) [inline python/ray/tune/track/project.py]: should go into separate PR

File: python/ray/tune/track/sync.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
+from __future__ import absolute_import
```
Comment (richardliaw) [inline python/ray/tune/track/sync.py]: cut from (tune/log_sync)

File: python/ray/tune/track/trial.py (hunk 4 above)
Comment (richardliaw) [inline python/ray/tune/track/trial.py]: TrialContext

File: python/ray/tune/track/trials.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
+"""
```
Comment (richardliaw) [inline python/ray/tune/track/trials.py]: same thing as `tune` cli

Comment (richardliaw) [reply python/ray/tune/track/trials.py]: cut :)

File: python/ray/tune/track/__init__.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
+import pickle
+
+from .trial import Trial
+from .project import Project
+from .log import debug
+from .convenience import absl_flags
```
Comment (noahgolmant) [inline python/ray/tune/track/__init__.py]: cut absl integration

File: python/ray/tune/track/__init__.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    Cleans up the trial and removes it from the global context.
+    """
+    global _trial  # pylint: disable=global-statement
+    if not _trial:
+        raise ValueError("Tried to stop trial, but no trial exists")
+    _trial.close()
+    _trial = None
+
+
+def save(obj, obj_name, iteration=None, save_fn=pickle.dump, **kwargs):
```
Comment (noahgolmant) [inline python/ray/tune/track/__init__.py]: write save/load/artifact test cases

File: python/ray/tune/track/__init__.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    """ Applies Trial.save to the trial in the current context """
+    return _trial.save(obj=obj, obj_name=obj_name, iteration=iteration,
+                       save_fn=save_fn, **kwargs)
+
+
+def metric(*, iteration=None, **kwargs):
```
Comment (noahgolmant) [inline python/ray/tune/track/__init__.py]: make python2-compatible

File: python/ray/tune/tests/test_track.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        def testme(config):
+            for i in range(config["iters"]):
+                track.metric(iteration=i, **{"hi": "test"})
+
+        trials = tune.run(testme, config={"iters": 5})
+        trial_res = trials[0].last_result
+        self.assertTrue(trial_res["hi"], "test")
+        self.assertTrue(trial_res["training_iteration"], 5)
+
+    # def testLocalMetrics(self):
```
Comment (richardliaw) [inline python/ray/tune/tests/test_track.py]: @noahgolmant we should figure out what to do for this one

File: python/ray/tune/track/__init__.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    return _session.metric(iteration=iteration, **kwargs)
+
+
+def trial_dir():
+    """Returns the directory where trial results are saved.
+
+    This includes json data containing the session's parameters and metrics.
+    """
+    _session = get_session()
+    return _session.trial_dir()
```
Comment (richardliaw) [inline python/ray/tune/track/__init__.py]: We need to test this

**Comment by richardliaw** [comment]
@ericl can you take a look? (I basically took ownership of dev on this PR so would be good to get another pair of eyes)

File: python/ray/tune/tests/test_track.py [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+    def testMetric(self):
+        track.init(trial_name="test_metric")
+        session = track.get_session()
+        for i in range(5):
+            track.metric(test=i)
+        result_path = os.path.join(session.logdir, EXPR_RESULT_FILE)
+        self.assertTrue(_check_json_val(result_path, "test", i))
+
+    def testRayOutput(self):
```
Comment (richardliaw) [inline python/ray/tune/tests/test_track.py]: This is the main usage showcase of track -> ray integration

File: python/ray/tune/track/__init__.py [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+def shutdown():
+    """Cleans up the trial and removes it from the global context."""
+
+    global _session
+    if _session:
+        _session.close()
+    _session = None
+
+
+def metric(iteration=None, **kwargs):
```
Comment (ericl) [inline python/ray/tune/track/__init__.py]: shouldn't this be `metrics`? And why special case iteration?

Comment (richardliaw) [reply python/ray/tune/track/__init__.py]: The goal to special case iteration is to allow you to insert multiple metrics to the same row from different parts of the codebase; I removed it for now and will do it in a later PR; also `s/metric/log`

**Comment by richardliaw** [comment]
ping @ericl

File: python/ray/tune/track/__init__.py [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+def shutdown():
+    """Cleans up the trial and removes it from the global context."""
+
+    global _session
+    if _session:
+        _session.close()
+    _session = None
+
+
+def log(iteration=None, **kwargs):
```
Comment (ericl) [inline python/ray/tune/track/__init__.py]: Why don't we remove iteration as a top-level arg?

**Review by ericl — APPROVED** [review]

**Comment by richardliaw** [comment]
TODO: fix for `signature` is https://pypi.org/project/funcsigs/ ?
